# Data Wrangling Capstone 1

 This portion of my Capstone Project is where I will perform several Data Wrangling Steps before beginning my analysis.
 
 My process will included the following:
 1. Importing the proper libraries
   + Pandas: To format the data into a Dataframe
   + Display: to display my dataframes in a nice format
   
 2. Importing the data
   + The data is in the form of a csv file so I will use Pandas.read_CSV function to import the CSV to a Dataframe
   
 3. Exploring the Dataframe
   + Using pandas.info() method to examine:
       + The number of rows and columns
       + The data type of each column
       + If there are any missing values
       + If there are any columns I might want to get rid of
 
 4. Cleaning the Dataframe
   + Change any needed data types
   + Remove any unneeded columns
   + Format any null/NaN values
       + Depending on the quantity and sustenance of the rull/Nan values will determine if I remove them completely, remove some of them, or fill them
   + Format the data into using "Tidy Data" methods:
       + Columns as variables
       + Rows as observations
       
 5. These are my intial steps and I am aware while exploring the data set these methods may alter. 
 

In [94]:
import pandas as pd
from IPython.display import display, Markdown 

#importing the data and changing object values into datetime values
df = pd.read_csv("houston_recovery_center_4_1_13_to_9_30_17.csv", \
                 parse_dates= ["timestamp","m_dd_yyyy_h_mm_rounded_to_15_2","los_length_of_stay_37_35_55"])
#copy of the original dataframe just in case I might need it later

original_df = df
df.info()
#pd.options.display.max_columns = None
display(df.head())

display(Markdown("## Counting the number of null values in each column"))
null_df = df.isnull().sum()

display(null_df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26830 entries, 0 to 26829
Data columns (total 31 columns):
hrcnewclientid                    26828 non-null float64
intake_day                        26828 non-null object
timestamp                         26828 non-null datetime64[ns]
shift                             26828 non-null float64
officer_badge                     24659 non-null object
district_beat                     22584 non-null object
law_enforcement                   25026 non-null object
transportation_in                 26828 non-null object
source                            26828 non-null object
substance                         26308 non-null object
drug_of_choice                    8176 non-null object
sex                               26826 non-null object
age                               26825 non-null float64
living_status                     26087 non-null object
veteran                           25073 non-null object
primary_care_doctor               23137 n

hrcnewclientid intake_day           timestamp  shift officer_badge  \
0     199007050.0        Wed 2013-04-10 13:15:00    1.0          5722   
1     148532038.0        Fri 2013-04-12 19:00:00    2.0          5820   
2     171756546.0        Fri 2013-04-12 20:45:00    2.0          3999   
3     199110127.0        Sat 2013-04-13 22:45:00    2.0          8725   
4     179651792.0        Sat 2013-04-13 15:15:00    2.0          4614   

  district_beat law_enforcement transportation_in source substance  \
0          1A10             HPD                LE     LE   Alcohol   
1          1A10             HPD                LE     LE       PCP   
2          1A10             HPD                LE     LE   Alcohol   
3          1A10             HPD                LE     LE   Alcohol   
4          3B10             HPD                LE     LE   Alcohol   

  drug_of_choice sex   age living_status veteran primary_care_doctor  \
0            NaN   M  51.0      Homeless      No                 NaN   
1            NaN   M  31.0   Private Res      No                 NaN   
2            NaN   M  24.0   Private Res     Yes                 NaN   
3            NaN   M  20.0   Private Res      No                 NaN   
4            NaN   M  51.0      Homeless      No                 NaN   

  insurance er_last_90_days_y_n er_90_days_alc_sub er_past_year_y_n  \
0       NaN                 NaN                NaN              NaN   
1       NaN                  No                 No               No   
2       NaN                  No                 No               No   
3       NaN                 Yes                Yes              Yes   
4       NaN                 NaN                NaN              NaN   

  er_year_alc_sub_y_n mental_health criminal_justice_history low_income_y_n  \
0                 NaN           NaN                      NaN            NaN   
1                  No            No                      Yes            NaN   
2                  No         Other                      Yes            NaN   
3                 Yes      Multiple                       No            NaN   
4                 NaN            No                      Yes            NaN   

  program a_plan_signed exit_complete req_for_service_forms  \
0    PART            No            No                    No   
1      No           Yes           Yes                    No   
2      No           Yes            No                    No   
3      No           Yes           Yes                    No   
4      No           Yes            No                    No   

  m_dd_yyyy_h_mm_rounded_to_15_2 los_length_of_stay_37_35_55    triage  
0            2013-04-10 13:45:00                     0:30:00       NaN  
1            2013-04-12 21:00:00                     2:00:00  Hospital  
2            2013-04-13 09:00:00                    12:15:00       NaN  
3            2013-04-14 05:00:00                     6:15:00       NaN  
4            2013-04-15 06:30:00                    39:15:00       NaN

## Counting the number of null values in each column

hrcnewclientid                        2
intake_day                            2
timestamp                             2
shift                                 2
officer_badge                      2171
district_beat                      4246
law_enforcement                    1804
transportation_in                     2
source                                2
substance                           522
drug_of_choice                    18654
sex                                   4
age                                   5
living_status                       743
veteran                            1757
primary_care_doctor                3693
insurance                          5325
er_last_90_days_y_n                3849
er_90_days_alc_sub                 3945
er_past_year_y_n                   3862
er_year_alc_sub_y_n                4042
mental_health                      3853
criminal_justice_history           3705
low_income_y_n                    11825
program                             706


## Exploring the Dataframe
   
   ### Using pandas.info() method to examine
   + **The number of rows and columns**
       + There are **31 columns** and **26830 rows**
   + **The data type of each column**
       + Original dataframe = 3 columns of floats and 28 columns of strings
       + Parsed dataframe = 3 columns of floats 2 columns of datetime objects and 26 columns of string objects. 
   + **If there are any missing values**
       + Out of the 31 columns 21 of those columns have a count of missing values over 500
       + *At this time I am not quite sure what I want to do with the missing data points. I will have to go column by column to determine this*
       + As I was going column by column I noticed the extra amount of time this was taking and if this dataset had double the columns it does now, this might take double the time. Normally I would've 
           + Separated the columns by quantitative and qualitative values
           + Filled the qualitative values with "Not Provided"
           + Filled the quantitative values with either 0, the mean of the columns values, or bfilled/ffilled the column based on a pattern 
           + **But** all the quantitative values do not have more than 10 null values so I'll fill all Nan values as "Not Provided" to make it easier for the average person to understand what's going on with missing data 
   + **If there are any columns I might want to get rid of**
       + I do not want to get rid of any columns at this current time because it can all tell a story. 
           
           



## Cleaning the Dataframe

+ **Change any needed data types**:

 *I completed this in my import statement, but "los_length_of_stay_37_35_55" remained a string object so I'll leave as is for 
    now*
    
    + timestamp from string to datetime
    + m_dd_yyyy_h_mm_rounded_to_15_2 from string to datetime
    + los_length_of_stay_37_35_55 from string to datetime
    
+ **Remove any unneeded columns**:
    + I will keep all the columns for now because I don't want to lose any data
    + Once the dataframe is cleaned properly I will just extract the columns I need for certain analysis 
----
+ **Format any null/NaN values**
    + I will fill all NaN values as "Not Provided"
----
+ **Format the data into using "Tidy Data" methods**:
    + Columns as variables
    + Rows as observations
    + **I believe the dataset is already tidy but I will check with my mentor**



In [92]:
display(Markdown("# Formating any null/NaN values"))

# Filling all the null values as not provided
filleddf = df.fillna(value = "Not Provided")
display(filleddf.head(15))

# Formating any null/NaN values

hrcnewclientid intake_day            timestamp shift officer_badge  \
0     1.99007e+08        Wed  2013-04-10 13:15:00     1          5722   
1     1.48532e+08        Fri  2013-04-12 19:00:00     2          5820   
2     1.71757e+08        Fri  2013-04-12 20:45:00     2          3999   
3      1.9911e+08        Sat  2013-04-13 22:45:00     2          8725   
4     1.79652e+08        Sat  2013-04-13 15:15:00     2          4614   
5     1.84572e+08       Tues  2013-04-16 02:45:00     3        119712   
6     1.34011e+08        Wed  2013-04-17 01:00:00     3          2559   
7     1.22436e+08       Thur  2013-04-18 00:30:00     3          8435   
8     1.86308e+08       Thur  2013-04-18 23:00:00     3          7209   
9     1.39871e+08        Fri  2013-04-19 03:45:00     3          3653   
10    1.83954e+08        Fri  2013-04-19 17:00:00     2          2085   
11    1.34903e+08        Fri  2013-04-19 17:15:00     2          1836   
12    1.07978e+08        Sat  2013-04-20 02:00:00     3          8725   
13    1.94873e+08        Sat  2013-04-20 23:45:00     3          8312   
14    1.58582e+08        Sun  2013-04-21 01:45:00     3          7867   

   district_beat law_enforcement transportation_in source substance  \
0           1A10             HPD                LE     LE   Alcohol   
1           1A10             HPD                LE     LE       PCP   
2           1A10             HPD                LE     LE   Alcohol   
3           1A10             HPD                LE     LE   Alcohol   
4           3B10             HPD                LE     LE   Alcohol   
5          10H50             HPD                LE     LE       PCP   
6           1A10             HPD                LE     LE     Other   
7             2A             HPD                LE     LE       PCP   
8           9C40             HPD                LE     LE   Alcohol   
9           1A10             HPD                LE     LE     Other   
10          1A10             HPD                LE     LE   Alcohol   
11  Not Provided             HPD                LE     LE   Alcohol   
12          1A10             HPD                LE     LE   Alcohol   
13         10H20             HPD                LE     LE   Alcohol   
14          3B10             HPD                LE     LE   Alcohol   

   drug_of_choice sex age living_status       veteran primary_care_doctor  \
0    Not Provided   M  51      Homeless            No        Not Provided   
1    Not Provided   M  31   Private Res            No        Not Provided   
2    Not Provided   M  24   Private Res           Yes        Not Provided   
3    Not Provided   M  20   Private Res            No        Not Provided   
4    Not Provided   M  51      Homeless            No        Not Provided   
5    Not Provided   M  37   Private Res            No        Not Provided   
6    Not Provided   M  46   Private Res            No        Not Provided   
7    Not Provided   M  60      Homeless            No        Not Provided   
8    Not Provided   M  42      Homeless            No        Not Provided   
9    Not Provided   F  19   Private Res  Not Provided        Not Provided   
10   Not Provided   F  58   Private Res  Not Provided        Not Provided   
11   Not Provided   M  51      Homeless            No        Not Provided   
12   Not Provided   M  19   Private Res  Not Provided        Not Provided   
13   Not Provided   T  22   Private Res            No        Not Provided   
14   Not Provided   M  48      Homeless            No        Not Provided   

       insurance er_last_90_days_y_n er_90_days_alc_sub er_past_year_y_n  \
0   Not Provided        Not Provided       Not Provided     Not Provided   
1   Not Provided                  No                 No               No   
2   Not Provided                  No                 No               No   
3   Not Provided                 Yes                Yes              Yes   
4   Not Provided        Not Provided       Not Provided     Not Provided   
5   Not 

## Not Using This

----



In [93]:
df.loc[(df["officer_badge"]).isnull(),"officer_badge"] = "Not Provided"
display(df.loc[:,["timestamp","shift","officer_badge"]].sort_values(by = "officer_badge").tail(10))
display(df.loc[(df["officer_badge"] == "Not Provided")])

timestamp  shift officer_badge
3411  2013-12-08 01:45:00    3.0           SGT
2554  2013-10-18 10:30:00    1.0           SGT
5227  2014-04-03 17:30:00    2.0           SGT
4951  2014-03-19 22:15:00    2.0           SGT
3580  2013-12-19 05:15:00    3.0           SGT
5212  2014-04-02 18:15:00    2.0           SGT
26537 2017-09-13 15:50:00    2.0            T6
26353 2017-09-04 21:58:00    2.0            TX
26352 2017-09-04 21:42:00    2.0            TX
24580 2017-05-13 00:15:00    3.0           Y79

hrcnewclientid intake_day           timestamp  shift officer_badge  \
161       196908168.0        Sun 2013-05-19 00:00:00    3.0  Not Provided   
265       188197782.0        Sun 2013-05-26 02:00:00    3.0  Not Provided   
527       165249523.0        Fri 2013-06-14 17:30:00    2.0  Not Provided   
534       132316822.0        Sat 2013-06-15 00:15:00    3.0  Not Provided   
726       181348479.0        Fri 2013-06-28 20:45:00    2.0  Not Provided   
735       158150370.0        Sat 2013-06-29 05:15:00    3.0  Not Provided   
874       178715261.0        Mon 2013-07-08 14:30:00    1.0  Not Provided   
879       180255260.0       Tues 2013-07-09 10:00:00    1.0  Not Provided   
1276      157421326.0       Tues 2013-08-06 09:15:00    1.0  Not Provided   
1341      136212989.0        Sun 2013-08-11 06:30:00    3.0  Not Provided   
1679      118739204.0        Mon 2013-09-02 16:00:00    2.0  Not Provided   
1773      170580230.0        Sat 2013-09-07 22:30:00    2.0  Not Provided   
1973      142675340.0       Thur 2013-09-19 14:30:00    1.0  Not Provided   
2403      102360880.0       Thur 2013-10-10 13:00:00    1.0  Not Provided   
2812      159348588.0        Wed 2013-10-30 15:15:00    2.0  Not Provided   
2904      170968931.0        Mon 2013-11-04 16:15:00    2.0  Not Provided   
3447      175516203.0       Tues 2013-12-10 17:30:00    2.0  Not Provided   
3630      177694446.0        Sun 2013-12-22 04:30:00    3.0  Not Provided   
3779      180988458.0       Thur 2014-01-02 06:30:00    3.0  Not Provided   
3788      129901275.0        Fri 2014-01-03 07:00:00    3.0  Not Provided   
3847      107480901.0        Mon 2014-01-06 15:30:00    2.0  Not Provided   
3890      195082347.0        Fri 2014-01-10 08:30:00    1.0  Not Provided   
3972      161641622.0        Wed 2014-01-15 14:45:00    1.0  Not Provided   
4000      142933745.0        Fri 2014-01-17 11:15:00    1.0  Not Provided   
4046      180964988.0       Tues 2014-01-21 09:30:00    1.0  Not Provided   
4159      142312103.0       Tues 2014-01-28 09:15:00    1.0  Not Provided   
4572      153668764.0       Tues 2014-02-25 13:15:00    1.0  Not Provided   
4664      111251839.0        Sun 2014-03-02 08:30:00    1.0  Not Provided   
4875      198300790.0        Sun 2014-03-16 01:30:00    3.0  Not Provided   
4927      168911205.0       Tues 2014-03-18 17:30:00    2.0  Not Provided   
...               ...        ...                 ...    ...           ...   
26651     108198943.0        Wed 2017-09-20 08:03:00    1.0  Not Provided   
26652     154907719.0        Wed 2017-09-20 11:46:00    1.0  Not Provided   
26653     100918272.0        Wed 2017-09-20 11:52:00    1.0  Not Provided   
26654     197699262.0        Wed 2017-09-20 11:57:00    1.0  Not Provided   
26668     117614878.0       Thur 2017-09-21 09:29:00    1.0  Not Provided   
26672     125655759.0       Thur 2017-09-21 00:48:00    3.0  Not Provided   
26673     174756532.0       Thur 2017-09-21 15:25:00    2.0  Not Provided   
26680     125655759.0       Thur 2017-09-21 21:51:00    2.0  Not Provided   
26689     186014799.0        Fri 2017-09-22 10:25:00    1.0  Not Provided   
26690     116643188.0        Fri 2017-09-22 11:34:00    1.0  Not Provided   
26691     190698975.0        Fri 2017-09-22 15:50:00    2.0  Not Provided   
26739     196477601.0        Mon 2017-09-25 11:03:00    1.0  Not Provided   
26741     117614878.0        Mon 2017-09-25 14:57:00    1.0  Not Provided   
26742     174756532.0        Mon 2017-09-25 15:50:00    2.0  Not Provided   
26758     171226178.0        Wed 2017-09-27 01:36:00    3.0  Not Provided   
26759     183974024.0        Wed 2017-09-27 01:36:00    3.0  Not Provided   
26761     152437204.0        Wed 2017-09-27 02:50:00    3.0  Not Provided   
26763     187562264.0        Wed 2017-09-27 08:53:00    1.0  Not Provided   
26767     110061905.0        Wed 2017-09-27 14:01:00    1.0  Not Provided   
26768     174542955.0        Wed 2017-09-27 14:01:00    1.0  Not Provided   
267

**Officer_badge**: 2171 null values
        + **Why are there null values**
            + Officers not providing badge number
            + Person didn't get picked up by an officer
        
        + **Pros of NaN officer badge**: 
            + I can see if there is a correlation between officers not providing badge numbers against another variable
            + I can isolate not having a badge number as a separate var to see if it's affecting attendance rates
        
        + **Cons of Nan officer_badge**:
            + Could not not be read by certain agg functions
            + Not knowing if there is a correlation between a particular officer and attendance rates 
            + The average person might not understand what an Nan value for an officer means. 
        + **Possible fillna values**
            + "Not Provided"